### PR TITLE
WIP : Add multiple docks

### PR DESCRIPTION
**Pull request by jdlom**
As discussed in #313.

I need to clean a little (mix of camelcase and snakecase for example, code organization...)

I can't figure out why I can't trigger the creation plot immediately after initializing a dock.

I'm waiting for your review.

### CODE REVIEW

**Comment by jdlom** [comment]
I'll add some unit tests too

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,6 @@ DataPlotly/i18n/
 
 .coverage
 .noseids
+
+# Vscode
+.vscode/*
```
Comment (ghtmtt) [inline .gitignore:20]:
```suggestion
.vscode/*

```

File: DataPlotly/data_plotly.py [hunk 2]
```diff
@@ -105,22 +112,52 @@ def initGui(self):
...
+        self.toolButton.setMenu(self.toolButtonMenu)
+        self.toolButton.setPopupMode(QToolButton.MenuButtonPopup)
+        self.toolBtnAction = self.iface.addToolBarWidget(self.toolButton)
+        self.toolButton.setDefaultAction(self.show_dock_action)
+
+
+        sub_actions = [
+            {
+                "text": self.tr("Add a new dock"),
+                "icon_path":icon,
```
Comment (ghtmtt) [inline DataPlotly/data_plotly.py:143]:
```suggestion
                "icon_path": icon,
```

File: DataPlotly/data_plotly.py [hunk 3]
```diff
@@ -105,22 +112,52 @@ def initGui(self):
...
+                "callback": self.dock_manager.addNewDockFromDlg,
+                "parent": self.iface.mainWindow(),
+                "toolbutton": self.toolButton,
+                "add_to_menu": False
+            },
+            {
+                "text": self.tr("Remove a dock"),
+                "icon_path":icon,
```
Comment (ghtmtt) [inline DataPlotly/data_plotly.py:151]:
```suggestion
                "icon_path": icon,
```

File: DataPlotly/gui/remove_dock_dlg.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        """Dialog to add new dock"""
+        super().__init__(parent)
+        self.setupUi(self)
+        dock_ids = [dock_id for dock_id in dock_widgets.keys() if dock_id != 'DataPlotly']
+        self.DockIdsComboBox.addItems(dock_ids)
+        for i, dock_id in enumerate(dock_ids):
+            self.DockIdsComboBox.setItemData(i, dock_widgets[dock_id].title, Qt.ToolTipRole)
+
+    def get_param(self):
+        return self.DockIdsComboBox.currentText()
```
Comment (ghtmtt) [inline DataPlotly/gui/remove_dock_dlg.py:23]:
```suggestion
        return self.DockIdsComboBox.currentText()
        
```

**Review by ghtmtt — COMMENTED** [review]
That is BRILLIANT, thanks @jdlom! I just have some really minor suggestions. Are you also able to add some tests?

**Comment by ghtmtt** [comment]
there are some `pylint` small problems. If you also have a look at that it would be great: https://github.com/ghtmtt/DataPlotly/actions/runs/4559132929/jobs/8137320295?pr=314

**Comment by jdlom** [comment]
@ghtmtt no problem to do that. I have no experience in unit tests or pylint (in python). Could you tell me where I can find some documentation ?

**Comment by ghtmtt** [comment]
You can have a look at the main repo folder to see existing tests:

https://github.com/ghtmtt/DataPlotly/tree/master/DataPlotly/test

to run on your local machine on Linux or hopefully also OSX (cannot be so helpful if you are on Windows). you have to tell where the QGIS output folder is (in the following example I'm running it on a compiled version of QGIS) and `run-env-linux.sh` is a bash file of the repository:

`source scripts/run-env-linux.sh /home/matteo/dev/qgis-master/output/`

and then use the  `Makefile` to run the tests:

`make test`

and

`make pylint`

unfortunately there isn't a lot of documentation around

**Comment by ghtmtt** [comment]
@jdlom can you please write me a provate email at info@faunalia.eu to discuss some detail?

**Comment by jdlom** [comment]
> @jdlom can you please write me a private email at [info@faunalia.eu](mailto:info@faunalia.eu) to discuss some detail?

It's done. 

I think I'll start adding the tests and fixing the pylint warnings this week end.

**Comment by jdlom** [comment]
@ghtmtt, I'll finish the day after tomorrow. Could you run the ci (pylint and test) ? I couldn't test on qgis 3.30 (I have setup my environment with miniconda  and it's a little bit tricky. Maybe a library problem. I couldn't figure out why running tests is working on qgis 3.18 but it's not on 3.30.

File: .gitignore [hunk 5]
```diff
@@ -18,3 +18,6 @@ DataPlotly/i18n/
 
 # Vscode
 .vscode/*
+
+# Test
+DataPlotly/test/*.qgs~
```
Comment (ghtmtt) [inline .gitignore:23]:
```suggestion
*.qgs~
```

let's ignore all the project backup files

File: DataPlotly/test/test_dock_manager.py [hunk 6]
```diff
@@ -137,10 +154,33 @@ def test_add_docks_from_project(self):
...
+        state, _, _ = validator.validate('NewDockId', None)
+        self.assertEqual(
+            QValidator.Acceptable,
+            state
+        )
+        # Dockid can not be empty / No underscore / Not already taken
+        for bad_dock_id in ['', 'with_underscore', 'DataPlotly2']:
+            state, _, _ = validator.validate(bad_dock_id, None)
+            self.assertEqual(QValidator.Intermediate, state)
+
```
Comment (ghtmtt) [inline DataPlotly/test/test_dock_manager.py:183]:
```suggestion
            self.assertEqual(QValidator.Intermediate, state)
```

**Review by ghtmtt — COMMENTED** [review]
Thanks for taking care of the changes of the syntax not related to you PR! I just added 2 minor suggestions, I've accepted them in order to run the tests.

**Comment by ghtmtt** [comment]
@jdlom thanks for this amazing contribution!

**Pull request merged**